File: Perl/examples/url_utils_example.pl
#!/usr/bin/perl
# URL Utilities Example
# 展示 URL 工具模块的各种用法

use strict;
use warnings;
use utf8;
use FindBin;
use lib "$FindBin::Bin/../url_utils";
use mod;

print "=" x 60 . "\n";
print "URL Utilities Example - Perl\n";
print "=" x 60 . "\n\n";

# ==================== 1. URL 编码与解码 ====================
print "1. URL 编码与解码\n";
print "-" x 40 . "\n";

my $text = "Hello World! 你好世界 @#$%^&*()";
print "原始文本: $text\n";

my $encoded = mod::url_encode($text);
print "编码后:   $encoded\n";

my $decoded = mod::url_decode($encoded);
print "解码后:   $decoded\n";

# 组件编码（对保留字符也编码）
my $component = "a+b=c&d=e";
print "\n组件编码示例:\n";
print "原始:     $component\n";
print "编码后:   " . mod::url_encode_component($component) . "\n";

# ==================== 2. 查询字符串处理 ====================
print "\n\n2. 查询字符串处理\n";
print "-" x 40 . "\n";

# 解析查询字符串
my $query = "name=John%20Doe&age=30&city=New%20York&tag=perl&tag=programming";
print "查询字符串: $query\n\n";

my $params = mod::parse_query_string($query);
print "解析结果:\n";
foreach my $key (sort keys %$params) {
    my $value = $params->{$key};
    if (ref $value eq 'ARRAY') {
        print "  $key = [" . join(", ", @$value) . "]\n";
    } else {
        print "  $key = $value\n";
    }
}

# 构建查询字符串
print "\n构建查询字符串:\n";
my $new_params = {
    search => "perl tutorial",
    page   => 1,
    sort   => "relevance"
};
my $new_query = mod::build_query_string($new_params);
print "参数: { search => 'perl tutorial', page => 1, sort => 'relevance' }\n";
print "结果: $new_query\n";

# 排序构建
my $sorted_query = mod::build_query_string($new_params, { sort => 1 });
print "排序后: $sorted_query\n";

# ==================== 3. URL 解析与构建 ====================
print "\n\n3. URL 解析与构建\n";
print "-" x 40 . "\n";

# 解析复杂 URL
my $complex_url = 'https://user:password@api.example.com:8080/v1/users?active=true&role=admin#section2';
print "解析 URL: $complex_url\n\n";

my $parsed = mod::parse_url($complex_url);
print "组件分解:\n";
print "  协议:   $parsed->{scheme}\n"   if $parsed->{scheme};
print "  用户:   $parsed->{userinfo}\n" if $parsed->{userinfo};
print "  主机:   $parsed->{host}\n"     if $parsed->{host};
print "  端口:   $parsed->{port}\n"     if $parsed->{port};
print "  路径:   $parsed->{path}\n"     if $parsed->{path};
print "  查询:   $parsed->{query}\n"    if $parsed->{query};
print "  锚点:   $parsed->{fragment}\n" if $parsed->{fragment};

# 从组件构建 URL
print "\n从组件构建 URL:\n";
my $built_url = mod::build_url({
    scheme => 'https',
    host   => 'github.com',
    path   => '/ayukyo/alltoolkit',
    query  => { tab => 'readme', lang => 'perl' }
});
print "构建结果: $built_url\n";

# ==================== 4. URL 验证与工具 ====================
print "\n\n4. URL 验证与工具\n";
print "-" x 40 . "\n";

my @urls_to_test = (
    "https://example.com",
    "http://localhost:3000/api",
    "ftp://files.example.com",
    "not-a-valid-url",
    "just-some-text",
    "",
);

print "URL 验证测试:\n";
foreach my $url (@urls_to_test) {
    my $valid = mod::is_valid_url($url);
    my $status = $valid ? "✓ 有效" : "✗ 无效";
    print "  '$url' -> $status\n";
}

# 提取域名和路径
print "\n提取域名和路径:\n";
my $test_url = "https://www.perl.org/documentation/";
print "URL: $test_url\n";
print "  域名: " . mod::get_domain($test_url) . "\n";
print "  路径: " . mod::get_path($test_url) . "\n";

# ==================== 5. 参数操作 ====================
print "\n\n5. 参数操作\n";
print "-" x 40 . "\n";

# 添加参数
my $base_url = "https://api.example.com/search";
print "基础 URL: $base_url\n";

my $with_params = mod::add_query_params($base_url, {
    q     => "perl programming",
    page  => 1,
    limit => 10
});
print "添加参数后: $with_params\n";

# 继续添加参数
my $with_more = mod::add_query_params($with_params, { sort => "relevance" });
print "再添加 sort: $with_more\n";

# 移除参数
my $removed = mod::remove_query_params($with_more, ["page", "limit"]);
print "移除 page 和 limit: $removed\n";

# ==================== 6. URL 规范化 ====================
print "\n\n6. URL 规范化\n";
print "-" x 40 . "\n";

my @urls_to_normalize = (
    "HTTPS://EXAMPLE.COM:443/path",
    "http://Example.COM:80/Page",
    "https://example.com/./path/../other",
);

print "URL 规范化:\n";
foreach my $url (@urls_to_normalize) {
    my $normalized = mod::normalize_url($url);
    print "  原始: $url\n";
    print "  规范: $normalized\n\n";
}

# ==================== 7. 实用场景示例 ====================
print "\n7. 实用场景示例\n";
print "-" x 40 . "\n";

# 场景 1: 构建带分页的 API URL
print "场景 1: 构建分页 API URL\n";
sub build_api_url {
    my ($base, $endpoint, $params) = @_;
    return mod::build_url({
        scheme => 'https',
        host   => $base,
        path   => $endpoint,
        query  => $params
    });
}

my $api_url = build_api_url('api.example.com', '/v1/users', {
    page  => 2,
    limit => 20,
    sort  => 'created_at'
});
print "API URL: $api_url\n";

# 场景 2: 处理回调 URL
print "\n场景 2: 处理回调 URL\n";
my $callback = "https://myapp.com/callback?code=abc123&state=xyz&session=temp";
print "原始回调: $callback\n";

# 移除敏感参数
my $clean_callback = mod::remove_query_params($callback, ["session"]);
print "清理后:   $clean_callback\n";

# 场景 3: 验证和规范化用户输入的 URL
print "\n场景 3: 验证用户输入\n";
my $user_input = "HTTPS://Example.COM:443/Path";
if (mod::is_valid_url($user_input)) {
    my $canonical = mod::normalize_url($user_input);
    print "输入: $user_input\n";
    print "规范: $canonical\n";
} else {
    print "无效的 URL: $user_input\n";
}

print "\n" . "=" x 60 . "\n";
print "示例运行完成!\n";
print "=" x 60 . "\n";

File: Rust/examples/url_utils_example.rs
//! URL Utilities Example
//!
//! Demonstrates usage of the URL utilities module

// Include the module
#[path = "../url_utils/mod.rs"]
mod url_utils;

use url_utils::*;
use std::collections::HashMap;

fn main() {
    println!("========================================");
    println!("Rust URL Utilities - Usage Examples");
    println!("========================================\n");

    // Example 1: Parse a URL
    println!("Example 1: Parse a URL");
    println!("----------------------");
    let url_str = "https://user:pass@api.example.com:8080/v1/users?page=1&limit=10#details";
    println!("Input: {}", url_str);
    
    match parse_url(url_str) {
        Ok(url) => {
            println!("Scheme: {}", url.scheme);
            println!("Host: {}", url.host);
            println!("Port: {:?}", url.port);
            println!("Path: {}", url.path);
            println!("Query 'page': {:?}", url.get_param("page"));
            println!("Query 'limit': {:?}", url.get_param("limit"));
            println!("Fragment: {:?}", url.fragment);
            println!("Origin: {}", url.origin());
            println!("Is Secure: {}", url.is_secure());
        }
        Err(e) => println!("Error: {}", e),
    }
    println!();

    // Example 2: Build a URL using UrlBuilder
    println!("Example 2: Build URL with UrlBuilder");
    println!("------------------------------------");
    let url = UrlBuilder::new()
        .scheme("https")
        .host("api.example.com")
        .path("/v2/search")
        .query_param("q", "rust programming")
        .query_param("category", "tutorials")
        .query_param("page", "1")
        .fragment("results")
        .build();
    println!("Built URL: {}", url);
    println!();

    // Example 3: URL Encoding and Decoding
    println!("Example 3: URL Encoding/Decoding");
    println!("--------------------------------");
    let original = "Hello, World! How are you?";
    let encoded = url_encode(original);
    let decoded = url_decode(&encoded);
    println!("Original: {}", original);
    println!("Encoded:  {}", encoded);
    println!("Decoded:  {}", decoded);
    println!();

    // Example 4: Query String Manipulation
    println!("Example 4: Query String Manipulation");
    println!("------------------------------------");
    let query = "name=John&age=30&city=New%20York";
    println!("Query string: {}", query);
    let params = parse_query_string(query);
    println!("Parsed params:");
    for (key, value) in &params {
        println!("  {} = {}", key, value);
    }
    
    // Build new query string
    let mut new_params = HashMap::new();
    new_params.insert("language".to_string(), "rust".to_string());
    new_params.insert("version".to_string(), "1.70".to_string());
    let new_query = build_query_string(&new_params);
    println!("Built query: {}", new_query);
    println!();

    // Example 5: URL Validation
    println!("Example 5: URL Validation");
    println!("-------------------------");
    let urls = vec![
        "https://example.com",
        "http://localhost:3000",
        "/relative/path",
        "not a valid url",
        "",
        "ftp://files.example.com",
    ];
    for url in urls {
        let valid = is_valid_url(url);
        println!("  '{}' -> {}", url, if valid { "VALID" } else { "INVALID" });
    }
    println!();

    // Example 6: Join URLs
    println!("Example 6: Join URLs");
    println!("--------------------");
    let base = "https://api.example.com/v1";
    let paths = vec!["users", "posts", "comments"];
    for path in paths {
        match join_url(base, path) {
            Ok(url) => println!("  {} + {} = {}", base, path, url),
            Err(e) => println!("  Error: {}", e),
        }
    }
    println!();

    // Example 7: Extract Domain and Path
    println!("Example 7: Extract Domain and Path");
    println!("----------------------------------");
    let test_urls = vec![
        "https://www.rust-lang.org/learn",
        "https://github.com/rust-lang/rust",
        "https://crates.io/crates/url",
    ];
    for url in test_urls {
        println!("URL: {}", url);
        println!("  Domain: {:?}", get_domain(url));
        println!("  Path: {:?}", get_path(url));
    }
    println!();

    // Example 8: Modify URL Parameters
    println!("Example 8: Modify URL Parameters");
    println!("--------------------------------");
    let mut url = parse_url("https://api.example.com/items?page=1").unwrap();
    println!("Original URL: {}", url.to_string());
    
    url.set_param("page", "2");
    url.set_param("limit", "20");
    url.set_param("sort", "desc");
    println!("After adding params: {}", url.to_string());
    
    url.remove_param("limit");
    println!("After removing 'limit': {}", url.to_string());
    println!();

    // Example 9: Normalize URL
    println!("Example 9: Normalize URL");
    println!("------------------------");
    let messy_url = "HTTPS://Example.COM:443/path?query=value";
    println!("Original: {}", messy_url);
    match normalize_url(messy_url) {
        Ok(normalized) => println!("Normalized: {}", normalized),
        Err(e) => println!("Error: {}", e),
    }
    println!();

    // Example 10: Complex URL Building
    println!("Example 10: Complex URL Building");
    println!("--------------------------------");
    let api_url = UrlBuilder::new()
        .scheme("https")
        .username("api_user")
        .password("secret_key")
        .host("api.service.com")
        .port(8443)
        .path("/v3/resources")
        .query_param("format", "json")
        .query_param("include", "metadata,tags")
        .query_param("limit", "100")
        .fragment("response")
        .build();
    println!("Complex URL: {}", api_url);
    
    // Parse it back
    match parse_url(&api_url) {
        Ok(parsed) => {
            println!("Parsed back:");
            println!("  Username: {:?}", parsed.username);
            println!("  Host: {}", parsed.host);
            println!("  Port: {:?}", parsed.port);
            println!("  Query params count: {}", parsed.query.len());
        }
        Err(e) => println!("Parse error: {}", e),
    }

    println!("\n========================================");
    println!("Examples completed!");
    println!("========================================");
}
